sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+      // NB: Elasticsearch 5.x now provides the slice API.
+      JsonObject statsJson = getStats(false);
+      JsonObject indexStats =
+          statsJson.getAsJsonObject("indices").getAsJsonObject(
+              spec.getConnectionConfiguration().getIndex()).getAsJsonObject("primaries");
+      JsonObject store = indexStats.getAsJsonObject("store");
+      return store.getAsJsonPrimitive("size_in_bytes").getAsLong();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("shard", shardPreference));
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+      return false;
+    }
+
+    @Override
+    public BoundedReader<String> createReader(PipelineOptions options) throws IOException {
+      return new BoundedElasticsearchReader(this);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public Coder<String> getDefaultOutputCoder() {
+      return StringUtf8Coder.of();
+    }
+
+    private JsonObject getStats(boolean shardLevel) throws IOException {
+      RestClient restClient = spec.getConnectionConfiguration().createClient();
+      HashMap<String, String> params = new HashMap<>();
+      if (shardLevel) {
+        params.put("level", "shards");
+      }
+      String endpoint = String.format("/%s/_stats",
+                                      spec.getConnectionConfiguration
+                                          ().getIndex());
+      Response response =
+          restClient.performRequest("GET", endpoint, params, new BasicHeader("", ""));
+      restClient.close();
+      return parseResponse(response);
+
+    }
+
+  }
+
+  private static class BoundedElasticsearchReader extends BoundedSource.BoundedReader<String> {
+
+    private final BoundedElasticsearchSource source;
+
+    private RestClient restClient;
+    private String current;
+    private String scroll;
+
+    public BoundedElasticsearchReader(BoundedElasticsearchSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      restClient = source.spec.getConnectionConfiguration().createClient();
+
+      String query = source.spec.getQuery();
+      if (query == null) {
+        query = "{\n"
+            + "  \"query\": {\n"
+            + "    \"match_all\": {}\n"
+            + "  }\n"
+            + "}";
+      }
+
+      Response response;
+      String endPoint = String.format("/%s/%s/_search",
+                                      source.spec.getConnectionConfiguration().getIndex(),
+                                      source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive() != null ? source.spec
```

Comment:
This is unfortunate because it means the only proper way to set this value for the user is trial-and-error: Beam doesn't give any guarantees on how long it will take between successive advance() calls. Even if you assume that it's "as soon as possible", the user still has no way to predict a correct value in advance, because e.g. in case of any runner that supports fusion, the answer will be "however long it will take to execute the composition of DoFn's the runner decided to fuse on top of this read" and this can not be predicted by the user. It also goes against the "no knobs" spirit of Beam - things should "just work" without tweaking parameters like batch sizes, keepalive timeouts, etc.

However, I don't see a good way to get rid of this particular parameter (the best I can come up with is "materialize the PCollection so that there's no fusion going on and Beam reads documents as fast as possible, and cross fingers", but this has its downsides).

Just please document that users should change it ONLY if their pipeline is failing with this particular error (...quote the error that happens if keepalive expires...), and never for any other reason.

Same comment applies to batch sizes. Users shouldn't be tweaking them unless the pipeline behaves unacceptably for a reason that can only be fixed by changing batch sizes.